src/storage.rs

```diff
@@ -0,0 +1,135 @@
+//! Storage
+/// Traits to allow on and off board storage deivces to read and write data
+/// Allows for Read Only Memory as well as Read and Write Memory
+use nb;
+
+/// Address represents an unsigned int. This allows for devices that have bigger or smaller address spaces than the host.
+pub struct Address<U>(pub U);
+/// Address Offset represents an unsigned int that is used as an optional offset from the base address.
+pub struct AddressOffset<U>(pub U);
+
+use core::ops::Add;
+
+/// Implement add for the Address and AddressOffset Types.
+impl<'a, 'b, U> Add<&'b AddressOffset<U>> for &'a Address<U>
+where
+    U: Add<U, Output = U> + Copy,
+{
+    type Output = Address<U>;
+
+    fn add(self, other: &'b AddressOffset<U>) -> Address<U> {
+        Address(self.0 + other.0)
+    }
+}
+
+/// Page represents an unsigned int that is a Page ID in the device memory space.
+pub struct Page<U>(U);
```

Comment:
I am wondering about a conversion from `Page` to `Address` but I guess the best way would be adding a `From` implementation and that is something that needs to be done in the implementation and not nicely enforcible here.